# What is asked
On the expenses data-set there is a column called platform
1. Tell us which platform is the one collecting the most campaigns?
2. What is the less used platform? Make a campaigns bar plot with all platforms.
3. What platform is collecting the most money?
4. What is the platform collecting the less money? Make an expenses bar plot with all platforms.
5. Can you draw conclusions?

On the revenues data-set, 
1. select all the enterprises having at least a whole year (Spawning more than 365 days, starting on any day) of revenue **and** expenses. Select that whole year, what are the enterprises having the biggest revenue and the lowest revenue? 
2. For each one of the selected enterprises and the selected year, make a plot with two graphs, the aggregated **monthly** revenue and the aggregated **monthly** expenses for that year.

# Libraries

In [1]:
import pandas as pd
import numpy as np
import plotly.express as px
import seaborn as sns

# Expenses analysis

In [40]:
# advertisement expenses an enterprise pays for a given platform and the date it was done
expenses = pd.read_csv('expenses_clean.csv')
expenses = expenses[['enterprise_id', 'date', 'platform', 'amount', 'total_funding_amount']]
expenses.columns = ['expenses_enterprise_id', 'expenses_date', 'expenses_platform', 'expenses_amount', 'expenses_total_funding_amount']
expenses.sort_values(by=['expenses_enterprise_id', 'expenses_date',  'expenses_platform', 'expenses_amount'], inplace=True)
expenses.reset_index(drop=True, inplace=True)

In [51]:
expenses.tail()

expenses_enterprise_id expenses_date expenses_platform  \
11626     9148671555452718366    2022-03-01          facebook   
11627     9148671555452718366    2022-03-24          facebook   
11628     9148671555452718366    2022-03-24         pinterest   
11629     9148671555452718366    2022-03-24         pinterest   
11630     9148671555452718366    2022-03-26         pinterest   

       expenses_amount  expenses_total_funding_amount  
11626           750.00                        32000.0  
11627           320.18                        32000.0  
11628           269.96                        32000.0  
11629           428.91                        32000.0  
11630            99.95                        32000.0

In [73]:
expenses.shape

(11631, 5)

In [71]:
expenses.drop_duplicates().shape

(9783, 5)

The dataset is named clean and the numbers that follow seem to make sense overall, so we will consider the identical lines are not duplicates

## Tell us which platform is the one collecting the most campaigns?

In [79]:
len(expenses['expenses_enterprise_id'].unique())

93

We have 93 unique enterprises in this dataset

In [53]:
# Zoom in on one company's platform to explore the data
expenses[(expenses['expenses_enterprise_id']==9148671555452718366)
        & (expenses['expenses_platform']=='facebook')]

expenses_enterprise_id expenses_date expenses_platform  \
11598     9148671555452718366    2021-12-09          facebook   
11599     9148671555452718366    2021-12-12          facebook   
11600     9148671555452718366    2021-12-17          facebook   
11601     9148671555452718366    2021-12-23          facebook   
11602     9148671555452718366    2021-12-28          facebook   
11603     9148671555452718366    2022-01-01          facebook   
11604     9148671555452718366    2022-01-03          facebook   
11605     9148671555452718366    2022-01-07          facebook   
11607     9148671555452718366    2022-01-12          facebook   
11608     9148671555452718366    2022-01-16          facebook   
11609     9148671555452718366    2022-01-19          facebook   
11610     9148671555452718366    2022-01-22          facebook   
11611     9148671555452718366    2022-01-25          facebook   
11613     9148671555452718366    2022-01-27          facebook   
11614     9148671555452718366    2022-01-30          facebook   
11615     9148671555452718366    2022-02-02          facebook   
11617     9148671555452718366    2022-02-03          facebook   
11618     9148671555452718366    2022-02-05          facebook   
11619     9148671555452718366    2022-02-08          facebook   
11620     9148671555452718366    2022-02-12          facebook   
11621     9148671555452718366    2022-02-14          facebook   
11622     9148671555452718366    2022-02-16          facebook   
11623     9148671555452718366    2022-02-20          facebook   
11624     9148671555452718366    2022-02-23          facebook   
11625     9148671555452718366    2022-02-26          facebook   
11626     9148671555452718366    2022-03-01          facebook   
11627     9148671555452718366    2022-03-24          facebook   

       expenses_amount  expenses_total_funding_amount  
11598           750.00                        32000.0  
11599           750.00                        32000.0  
11600           749.99                        32000.0  
11601           750.00                        32000.0  
11602           750.00                        32000.0  
11603           750.00                        32000.0  
11604           143.34                        32000.0  
11605           749.81                        32000.0  
11607           749.99                        32000.0  
11608           750.00                        32000.0  
11609           750.00                        32000.0  
11610           750.00                        32000.0  
11611           750.00                        32000.0  
11613           750.00                        32000.0  
11614           750.00                        32000.0  
11615           750.00                        32000.0  
11617           150.76                        32000.0  
11618           743.52                        32000.0  
11619           750.00                        32000.0  
11620           750.00                        32000.0  
11621           750.00                        32000.0  
11622           750.00                        32000.0  
11623           750.00                        32000.0  
11624           750.00                        32000.0  
11625           750.00                        32000.0  
11626           750.00                        32000.0  
11627           320.18                        32000.0

In [45]:
# Zoom in on another company's platform to explore the data
expenses[(expenses['expenses_enterprise_id']==275660407194757655)
        & (expenses['expenses_platform']=='facebook')]

expenses_enterprise_id expenses_date expenses_platform  expenses_amount  \
0       275660407194757655    2021-10-05          facebook             2.00   
1       275660407194757655    2021-10-05          facebook             4.00   
2       275660407194757655    2021-10-05          facebook             6.00   
3       275660407194757655    2021-10-05          facebook             9.00   
4       275660407194757655    2021-10-05          facebook            12.00   
5       275660407194757655    2021-10-05          facebook            20.00   
6       275660407194757655    2021-10-05          facebook            30.00   
7       275660407194757655    2021-10-06          facebook            40.00   
8       275660407194757655    2021-10-06          facebook            70.00   
9       275660407194757655    2021-10-07          facebook           100.00   
10      275660407194757655    2021-10-09          facebook           150.00   
12      275660407194757655    2021-10-11          facebook           225.00   
13      275660407194757655    2021-10-12          facebook           174.98   
14      275660407194757655    2021-10-13          facebook           350.00   
15      275660407194757655    2021-10-15          facebook           500.00   
17      275660407194757655    2021-10-19          facebook           750.00   
19      275660407194757655    2021-10-24          facebook           750.00   
20      275660407194757655    2021-10-27          facebook           750.00   
22      275660407194757655    2021-10-30          facebook           750.00   
24      275660407194757655    2021-11-02          facebook           750.00   
26      275660407194757655    2021-11-04          facebook           750.00   
28      275660407194757655    2021-11-06          facebook           750.00   
30      275660407194757655    2021-11-07          facebook           750.00   
32      275660407194757655    2021-11-09          facebook           750.00   
33      275660407194757655    2021-11-10          facebook           750.00   
34      275660407194757655    2021-11-11          facebook           750.00   
36      275660407194757655    2021-11-12          facebook            90.44   
38      275660407194757655    2021-11-13          facebook           750.00   
40      275660407194757655    2021-11-14          facebook           750.00   
44      275660407194757655    2021-11-16          facebook           750.00   
46      275660407194757655    2021-11-17          facebook           739.10   
47      275660407194757655    2021-11-19          facebook           750.00   
48      275660407194757655    2021-11-20          facebook           750.00   
50      275660407194757655    2021-11-21          facebook           750.00   
51      275660407194757655    2021-11-22          facebook           750.00   
52      275660407194757655    2021-11-23          facebook           750.00   
54      275660407194757655    2021-11-25          facebook           750.00   
56      275660407194757655    2021-11-26          facebook           750.00   
57      275660407194757655    2021-11-27          facebook           750.00   
59      275660407194757655    2021-11-28          facebook           750.00   
63      275660407194757655    2021-12-01          facebook           750.00   
65      275660407194757655    2021-12-03          facebook           750.00   
68      275660407194757655    2021-12-06          facebook           750.00   
70      275660407194757655    2021-12-08          facebook           750.00   
71      275660407194757655    2021-12-10          facebook           750.00   
73      275660407194757655    2021-12-11          facebook            33.19   
74      275660407194757655    2021-12-13          facebook           750.00   
76      275660407194757655    2021-12-16          facebook           750.00   

    expenses_total_funding_amount  
0                         40000.0  
1                         40000.0  
2                       

As we do not have any indication regarding how many campaigns a company runs for a specific date,
we will be considering we have only one campaign running per company per platform. <br>
Indeed, although the amounts can be similar over multiple days,
the assignment.md file specifies: "expenses an enterprise pays for a given platform and the date it was done"
which means each row is a different expense, despite similar amounts. <br>

In that case, we currently have 245 campaigns running, an average of 2.6 campaigns per company and Facebook has the largest number of campaigns with 80 campaigns running

In [116]:
# Compute the number of campaigns per platform
expenses_campaign_ranking = expenses.groupby('expenses_platform', as_index=False)['expenses_enterprise_id'].agg('nunique').sort_values(by='expenses_enterprise_id', ascending=False)
expenses_campaign_ranking.columns = ['expenses_platform', 'number_of_campaigns']

In [92]:
expenses_campaign_ranking['number_of_campaigns'].sum()

245

In [95]:
expenses_campaign_ranking['number_of_campaigns'].sum()/len(expenses['expenses_enterprise_id'].unique())

2.6344086021505375

In [96]:
expenses_campaign_ranking

expenses_platform  number_of_campaigns
0          facebook                   80
1            google                   74
3             other                   43
4         pinterest                   20
6             stock                   13
2         microsoft                    8
5          snapchat                    7

## What is the less used platform? Make a campaigns bar plot with all platforms.


In [ ]:
# The less used platform here is Snapchat with only 7 campaigns running

In [97]:
fig = px.bar(expenses_campaign_ranking, x='expenses_platform', y='number_of_campaigns')
fig.show()

## What platform is collecting the most money?


In [98]:
expenses_amount_ranking = expenses.groupby('expenses_platform', as_index=False)['expenses_amount'].agg('sum').sort_values(by='expenses_amount', ascending=False)

In [99]:
expenses_amount_ranking

expenses_platform  expenses_amount
0          facebook       6138277.39
1            google       3004154.18
6             stock       1749221.66
3             other       1045335.71
5          snapchat        331474.74
4         pinterest        127941.49
2         microsoft         73076.71

The platform collecting the most money is Facebook with 6,138,277.39€. spent on advertising on the platform

## What is the platform collecting the less money? Make an expenses bar plot with all platforms.


In [100]:
fig = px.bar(expenses_amount_ranking, x='expenses_platform', y='expenses_amount')
fig.show()

## Can you draw conclusions?

In [114]:
# Compute the distribution of expenses per platform
expenses_amount_ranking['expenses_amount_pct'] = expenses_amount_ranking['expenses_amount']/expenses_amount_ranking['expenses_amount'].sum()*100
expenses_amount_ranking['expenses_amount_pct'] = expenses_amount_ranking['expenses_amount_pct'].round(2)

In [115]:
expenses_amount_ranking

expenses_platform  expenses_amount  expenses_amount_pct
0          facebook       6138277.39                49.23
1            google       3004154.18                24.09
6             stock       1749221.66                14.03
3             other       1045335.71                 8.38
5          snapchat        331474.74                 2.66
4         pinterest        127941.49                 1.03
2         microsoft         73076.71                 0.59

In [120]:
# Compute the distribution of the number of campaigns per platform
expenses_campaign_ranking['number_of_campaigns_pct'] = expenses_campaign_ranking['number_of_campaigns']/expenses_campaign_ranking['number_of_campaigns'].sum()*100
expenses_campaign_ranking['number_of_campaigns_pct'] = expenses_campaign_ranking['number_of_campaigns_pct'].round(2)

In [121]:
expenses_campaign_ranking

expenses_platform  number_of_campaigns  number_of_campaigns_pct
0          facebook                   80                    32.65
1            google                   74                    30.20
3             other                   43                    17.55
4         pinterest                   20                     8.16
6             stock                   13                     5.31
2         microsoft                    8                     3.27
5          snapchat                    7                     2.86

Facebook and Google seem to be the most popular advertising platforms: they gather 62% of the campaigns and 75% of the expenses of those. <br>
Although Google and Facebook are roughly equal in relative number of campaigns (respectively 30.2% and 32.65% of the campaigns), expenses are much higher on Facebook (respectively 49.23% and 24.09% of the expenses)

In [437]:
# Compare expenses with funding
expenses_funding = expenses.groupby(['expenses_enterprise_id', 'expenses_total_funding_amount'], as_index=False)['expenses_amount'].agg('sum')
expenses_funding['ratio'] = expenses_funding['expenses_amount']/expenses_funding['expenses_total_funding_amount']

In [444]:
expenses_funding.describe()

expenses_enterprise_id  expenses_total_funding_amount  expenses_amount      ratio
count            9.300000e+01                      93.000000        93.000000  93.000000
mean             4.671823e+18                  148451.612903    134080.450323   0.854801
std              2.645339e+18                  199514.384146    188320.470405   0.219508
min              2.756604e+17                   10000.000000       500.000000   0.005000
25%              2.753454e+18                   32000.000000     28287.040000   0.771811
50%              4.736953e+18                   80000.000000     60000.000000   0.966260
75%              7.017315e+18                  160000.000000    150090.010000   0.998540
max              9.148672e+18                 1000000.000000    962788.930000   1.024210

Most of the companies have used a large part of their funding (75% have used 77% or more of their funding)

# Revenue analysis

In [4]:
# Cash flow and dates a given enterprise has received
revenue = pd.read_csv('revenue_clean.csv')
revenue.columns = ['revenue_enterprise_id', 'revenue_day_date', 'revenue_amount']

In [32]:
revenue.head()

revenue_enterprise_id revenue_day_date  revenue_amount
0    8749991449499427134       2020-05-31          926.00
1    8749991449499427134       2020-06-07        25050.41
2    8749991449499427134       2020-06-14        33036.67
3    8749991449499427134       2020-06-21        29512.00
4    8749991449499427134       2020-06-28        21441.22

In [74]:
revenue.shape

(25216, 3)

In [75]:
revenue.drop_duplicates().shape

(25216, 3)

## select all the enterprises having at least a whole year (Spawning more than 365 days, starting on any day) of revenue **and** expenses. Select that whole year, what are the enterprises having the biggest revenue and the lowest revenue? 

### Merge the datasets

In [445]:
expenses.head()

expenses_enterprise_id expenses_date expenses_platform  expenses_amount  expenses_total_funding_amount
0      275660407194757655    2021-10-05          facebook              2.0                        40000.0
1      275660407194757655    2021-10-05          facebook              4.0                        40000.0
2      275660407194757655    2021-10-05          facebook              6.0                        40000.0
3      275660407194757655    2021-10-05          facebook              9.0                        40000.0
4      275660407194757655    2021-10-05          facebook             12.0                        40000.0

In [446]:
revenue.head()

revenue_enterprise_id revenue_day_date  revenue_amount
0    8749991449499427134       2020-05-31          926.00
1    8749991449499427134       2020-06-07        25050.41
2    8749991449499427134       2020-06-14        33036.67
3    8749991449499427134       2020-06-21        29512.00
4    8749991449499427134       2020-06-28        21441.22

In [161]:
revenue.shape

(25216, 3)

In [160]:
revenue[['revenue_enterprise_id', 'revenue_day_date']].drop_duplicates().shape

(25216, 2)

In [174]:
# Aggregate expenses to avoid duplicates when joining the tables
agg_expenses = expenses.groupby(['expenses_enterprise_id', 'expenses_date', 'expenses_total_funding_amount'], as_index=False)['expenses_amount'].agg('sum')

In [175]:
agg_expenses.shape

(7390, 4)

In [176]:
agg_expenses[['expenses_enterprise_id', 'expenses_date']].drop_duplicates()

expenses_enterprise_id expenses_date
0         275660407194757655    2021-10-05
1         275660407194757655    2021-10-06
2         275660407194757655    2021-10-07
3         275660407194757655    2021-10-09
4         275660407194757655    2021-10-10
...                      ...           ...
7385     9148671555452718366    2022-02-23
7386     9148671555452718366    2022-02-26
7387     9148671555452718366    2022-03-01
7388     9148671555452718366    2022-03-24
7389     9148671555452718366    2022-03-26

[7390 rows x 2 columns]

In [226]:
revenue

revenue_enterprise_id revenue_day_date  revenue_amount
0        8749991449499427134       2020-05-31          926.00
1        8749991449499427134       2020-06-07        25050.41
2        8749991449499427134       2020-06-14        33036.67
3        8749991449499427134       2020-06-21        29512.00
4        8749991449499427134       2020-06-28        21441.22
...                      ...              ...             ...
25211    4967883010146339226       2022-03-16          234.84
25212    4967883010146339226       2022-03-13          102.03
25213    4967883010146339226       2022-03-21          180.09
25214    4967883010146339226       2022-03-22           35.67
25215    4967883010146339226       2022-03-23           29.00

[25216 rows x 3 columns]

In [257]:
# Merge the two datasets
revenue_and_expenses = pd.merge(agg_expenses, revenue, 
                                how='outer', 
                                left_on=['expenses_enterprise_id', 'expenses_date'],
                                right_on=['revenue_enterprise_id', 'revenue_day_date'])

# Fill NAs
revenue_and_expenses['expenses_enterprise_id'].fillna(revenue_and_expenses['revenue_enterprise_id'], inplace=True)
revenue_and_expenses['expenses_date'].fillna(revenue_and_expenses['revenue_day_date'], inplace=True)

# Select the relevant columns
revenue_and_expenses = revenue_and_expenses[['expenses_enterprise_id', 
                                             'expenses_date', 
                                             'expenses_amount', 
                                             'revenue_amount']]
# Filter the columns which are useful
revenue_and_expenses.columns = ['enterprise_id', 
                                'date', 
                                'expenses_amount',
                                'revenue_amount']

# Fill NAs
revenue_and_expenses['expenses_amount'].fillna(0, inplace=True)
revenue_and_expenses['revenue_amount'].fillna(0, inplace=True)

In [ ]:
# Checks

In [258]:
revenue.shape

(25216, 3)

In [259]:
expenses.shape

(11631, 5)

In [260]:
revenue_and_expenses.shape

(26095, 4)

In [261]:
revenue_and_expenses[revenue_and_expenses['enterprise_id']==275660407194757664.0]

enterprise_id        date  expenses_amount  revenue_amount
0       2.756604e+17  2021-10-05             83.0          160.00
1       2.756604e+17  2021-10-06            110.0          329.50
2       2.756604e+17  2021-10-07            100.0          118.00
3       2.756604e+17  2021-10-09            150.0          976.60
4       2.756604e+17  2021-10-10            500.0           99.00
...              ...         ...              ...             ...
16850   2.756604e+17  2022-03-08              0.0          387.50
16851   2.756604e+17  2022-03-21              0.0          646.50
16852   2.756604e+17  2022-03-22              0.0          297.00
16853   2.756604e+17  2022-03-19              0.0          596.25
16854   2.756604e+17  2022-03-23              0.0          269.25

[218 rows x 4 columns]

In [262]:
revenue[revenue['revenue_enterprise_id']==275660407194757664.0]

revenue_enterprise_id revenue_day_date  revenue_amount
12375     275660407194757655       2022-01-05          170.50
12376     275660407194757655       2022-01-06           34.50
12377     275660407194757655       2021-08-06          517.30
12378     275660407194757655       2021-08-08          159.50
12379     275660407194757655       2021-08-09          216.50
...                      ...              ...             ...
12585     275660407194757655       2022-03-08          387.50
12586     275660407194757655       2022-03-21          646.50
12587     275660407194757655       2022-03-22          297.00
12588     275660407194757655       2022-03-19          596.25
12589     275660407194757655       2022-03-23          269.25

[215 rows x 3 columns]

In [263]:
expenses[expenses['expenses_enterprise_id']==275660407194757664.0]

expenses_enterprise_id expenses_date expenses_platform  expenses_amount  \
0       275660407194757655    2021-10-05          facebook             2.00   
1       275660407194757655    2021-10-05          facebook             4.00   
2       275660407194757655    2021-10-05          facebook             6.00   
3       275660407194757655    2021-10-05          facebook             9.00   
4       275660407194757655    2021-10-05          facebook            12.00   
..                     ...           ...               ...              ...   
72      275660407194757655    2021-12-10            google           500.00   
73      275660407194757655    2021-12-11          facebook            33.19   
74      275660407194757655    2021-12-13          facebook           750.00   
75      275660407194757655    2021-12-13            google           500.00   
76      275660407194757655    2021-12-16          facebook           750.00   

    expenses_total_funding_amount  
0                         40000.0  
1                         40000.0  
2                         40000.0  
3                         40000.0  
4                         40000.0  
..                            ...  
72                        40000.0  
73                        40000.0  
74                        40000.0  
75                        40000.0  
76                        40000.0  

[77 rows x 5 columns]

In [270]:
# Filter on companies with at least 365 days of revenue and expenses
enterprises_filter = revenue_and_expenses.groupby('enterprise_id', as_index=False)['date'].agg('count').sort_values(by='date', ascending=False)
enterprises_filter.columns = ['enterprise_id', 'number_of_days']
enterprises_filter = tuple(enterprises_filter[enterprises_filter['number_of_days']>=365]['enterprise_id'])

In [277]:
# Apply the filter to the table and rank it to select a period of a year
filtered_revenue_and_expenses = revenue_and_expenses[revenue_and_expenses['enterprise_id'].isin(enterprises_filter)]
filtered_revenue_and_expenses.reset_index(drop=True, inplace=True)
filtered_revenue_and_expenses['rank'] = filtered_revenue_and_expenses.groupby(['enterprise_id'])['date'].cumcount()+1

In [306]:
# Select the period of a year
filtered_revenue_and_expenses_year = filtered_revenue_and_expenses[filtered_revenue_and_expenses['rank']<=365]
filtered_revenue_and_expenses_year.drop('rank', axis=1, inplace=True)

/usr/local/lib/python3.9/site-packages/pandas/core/frame.py:4308: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [336]:
# Rank the enterprises by revenue
revenue_rank = filtered_revenue_and_expenses_year.groupby('enterprise_id', as_index=False)['revenue_amount'].agg('sum').sort_values(by='revenue_amount', ascending=False)
revenue_rank.reset_index(drop=True, inplace=True)

In [357]:
revenue_rank['enterprise_id'] = revenue_rank['enterprise_id'].astype(int)

In [362]:
revenue_rank.head(1)

enterprise_id  revenue_amount
0  6775957183556624384      9833412.78

In [364]:
revenue_rank.tail(1)

enterprise_id  revenue_amount
30  5469371784944668672        97400.56

The enterprise with the largest revenue is the enterprise 6775957183556624384 with a revenue of 9,833,412.78 euros. The one with the lowest revenue is the enterprise 5469371784944668672 with a revenue of 97,400.56 euros.

## For each one of the selected enterprises and the selected year, make a plot with two graphs, the aggregated **monthly** revenue and the aggregated **monthly** expenses for that year.

In [365]:
filtered_revenue_and_expenses_year

enterprise_id        date  expenses_amount  revenue_amount
0       4.188632e+17  2021-05-06            500.0         1353.30
1       4.188632e+17  2021-05-07            750.0         1456.80
2       4.188632e+17  2021-05-08            750.0         1479.10
3       4.188632e+17  2021-05-09           1250.0         1959.90
4       4.188632e+17  2021-05-11            750.0         1941.70
...              ...         ...              ...             ...
14495   4.736953e+18  2021-12-31              0.0          893.35
14496   4.736953e+18  2022-01-01              0.0         1012.19
14497   4.736953e+18  2022-01-04              0.0         1741.25
14498   4.736953e+18  2022-01-15              0.0         2467.17
14499   4.736953e+18  2022-02-05              0.0         3151.22

[11315 rows x 4 columns]

In [371]:
# Convert the date column (str) to a datetime
filtered_revenue_and_expenses_year['date'] = pd.to_datetime(filtered_revenue_and_expenses_year['date'])

<ipython-input-371-a2191e2ba667>:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [393]:
# Select the relevant enterprise
max_enterprise = filtered_revenue_and_expenses_year[filtered_revenue_and_expenses_year['enterprise_id']==6775957183556624384].drop('enterprise_id', axis=1)
min_enterprise = filtered_revenue_and_expenses_year[filtered_revenue_and_expenses_year['enterprise_id']==6775957183556624384].drop('enterprise_id', axis=1)

In [431]:
## Prepare the data to be plotted for each enterprise
# enterprise with the maximum revenue
max_plot = max_enterprise.resample('M', on='date').sum()
max_plot['agg_expenses'] = max_plot['expenses_amount'].cumsum()
max_plot['agg_revenue'] = max_plot['revenue_amount'].cumsum()
max_plot.reset_index(inplace=True)
max_plot = max_plot[['date', 'agg_expenses', 'agg_revenue']]
max_plot = max_plot.melt(id_vars=['date'], 
                         var_name='expenses', 
                         value_name='amount')

# enterprise with the minimum revenue
min_plot = min_enterprise.resample('M', on='date').sum()
min_plot['agg_expenses'] = min_plot['expenses_amount'].cumsum()
min_plot['agg_revenue'] = min_plot['revenue_amount'].cumsum()
min_plot.reset_index(inplace=True)
min_plot = min_plot[['date', 'agg_expenses', 'agg_revenue']]
min_plot = min_plot.melt(id_vars=['date'], 
                         var_name='expenses', 
                         value_name='amount')

In [432]:
# Plot for the company with the maximum revenue
fig = px.line(max_plot, x="date", y="amount", color='expenses', title='Expenses')
fig.show()

In [433]:
# Plot for the company with the minimum revenue
fig = px.line(min_plot, x="date", y="amount", color='expenses', title='Expenses')
fig.show()